core/source_map.rs

```diff
@@ -89,13 +95,93 @@ impl SourceMapper {
     std::mem::take(&mut self.ext_source_maps)
   }
 
+  pub fn apply_source_map_from_module_map(
+    &mut self,
+    scope: &mut v8::HandleScope,
+    file_name: &str,
+    line_number: u32,
+    column_number: u32,
+  ) -> Option<SourceMapApplication> {
+    let module_map_rc = JsRealm::module_map_from(scope);
+    let id = module_map_rc.get_id(file_name, RequestedModuleType::None)?;
+
+    let module_handle = module_map_rc.get_handle(id).unwrap();
+    let module = v8::Local::new(scope, module_handle);
+    let unbound_module_script = module.get_unbound_module_script(scope);
+    let maybe_source_mapping_url =
+      unbound_module_script.get_source_mapping_url(scope);
+
+    // TODO(bartlomieju): This should be the last fallback and it's only useful
+    // for eval - probably for `Deno.core.evalContext()`.
+    // let maybe_source_url = unbound_module_script
+    //   .get_source_url(scope)
+    //   .to_rust_string_lossy(scope);
+    // eprintln!("maybe source url {}", maybe_source_url);
+
+    if !maybe_source_mapping_url.is_string() {
+      return None;
+    }
+
+    let source_map_string =
+      maybe_source_mapping_url.to_rust_string_lossy(scope);
+    // eprintln!("maybe source mapping url {}", source_map_string);
+
+    // TODO(bartlomieju): this is a fast path - if it fails, we should try to parse
+    // the URL (or resolve it from the current file being mapped) and fallback to
+    // acquiring a source map from that URL. In Deno we might want to apply permissions
+    // checks for fetching the map.
+    let source_map = if let Some(b64) =
+      source_map_string.strip_prefix("data:application/json;base64,")
+    {
+      let decoded_b64 = BASE64_STANDARD.decode(b64).ok()?;
+      // eprintln!(
+      //   "source map {:?}",
+      //   String::from_utf8(decoded_b64.clone()).unwrap()
+      // );
+      SourceMap::from_slice(&decoded_b64).ok()?
+    } else {
+      let url = match resolve_import(&source_map_string, file_name) {
+        Ok(url) => Some(url),
+        Err(err) => match err {
+          ModuleResolutionError::ImportPrefixMissing(_, _) => {
+            resolve_import(&format!("./{}", source_map_string), file_name).ok()
+          }
+          _ => None,
+        },
+      };
+      // eprintln!(
+      //   "source map url sourceMappingURL={} file_name={} url={}",
+      //   source_map_string,
+      //   file_name,
+      //   url.as_ref().map(|s| s.as_str()).unwrap_or_default()
+      // );
+      let url = url?;
+      if url.scheme() != "file" {
+        return None;
+      }
+      let contents = module_map_rc
+        .loader
+        .borrow()
+        .get_source_map(url.to_file_path().ok()?.to_str()?)?;
```

Comment:
This should be renamed to `get_source_map_file()` and do permission checks in the CLI